include/gabia/secure_socket.hpp

```diff
@@ -185,47 +188,61 @@ class secure_socket {
+            : state_ptr(std::forward<DeducedHandler>(handler), socket,
+                        sequence) {}
 
         void operator()(beast::error_code error, size_t n) {
             CORO_REENTER(*this) {
                 static_cast<void>(n);
-                socket.encrypt_pdu(sequence, pdu);
-                CORO_YIELD asio::async_write(socket.next_layer(),
-                                             asio::buffer(pdu), *this);
-                handler(error, asio::buffer_size(sequence));
+
+                state_ptr->socket.encrypt_pdu(state_ptr->sequence,
+                                              state_ptr->pdu);
+                CORO_YIELD asio::async_write(state_ptr->socket.next_layer(),
+                                             asio::buffer(state_ptr->pdu),
+                                             *this);
+                auto n = asio::buffer_size(state_ptr->sequence);
+                state_ptr.invoke(error, n);
             }
         }
 
         friend void* asio_handler_allocate(std::size_t size,
                                            async_write_some_op* op) {
             using boost::asio::asio_handler_allocate;
-            return asio_handler_allocate(size, std::addressof(op->handler));
+            return asio_handler_allocate(
+                size, std::addressof(op->state_ptr.handler()));
         }
 
         friend void asio_handler_deallocate(void* p, std::size_t size,
                                             async_write_some_op* op) {
             using boost::asio::asio_handler_deallocate;
-            asio_handler_deallocate(p, size, std::addressof(op->handler));
+            asio_handler_deallocate(p, size,
+                                    std::addressof(op->state_ptr.handler()));
         }
 
         friend bool asio_handler_is_continuation(async_write_some_op* op) {
             using boost::asio::asio_handler_is_continuation;
-            return asio_handler_is_continuation(std::addressof(op->handler));
+            return asio_handler_is_continuation(
+                std::addressof(op->state_ptr.handler()));
         }
 
         template <class Function>
         friend void asio_handler_invoke(Function&& f, async_write_some_op* op) {
             using boost::asio::asio_handler_invoke;
-            asio_handler_invoke(f, std::addressof(op->handler));
+            asio_handler_invoke(f, std::addressof(op->state_ptr.handler()));
         }
 
-        secure_socket& socket;
-        ConstBufferSequence sequence;
-        Handler handler;
-        std::vector<gsl::byte> pdu;
+        struct state {
+            template <typename DeducedHandler>
+            state(DeducedHandler&&, secure_socket& socket,
+                  const ConstBufferSequence& sequence)
+                : socket{socket}, sequence{sequence} {}
+
+            secure_socket& socket;
+            ConstBufferSequence sequence;
+            std::vector<gsl::byte> pdu;
```

Comment:
Nope, don't bother with the allocator in the synchronous case because you won't get the benefit. Although if you care about synchronous performance you should use a stack buffer instead of a vector.